File: internal/mechanics/leveling.go
package mechanics

import (
	"github.com/slycrel/slycrel/internal/model"
)

// NextLevelUp calculates the total experience required for a given level.
// Original: NextLevelUp() in Sly_CommonGuildUnit.cp
// Formula: f(x) = LevelUpDifficulty * x^3 + f(x-1), base = 50
func NextLevelUp(level int, difficulty int) int64 {
	if level <= 1 {
		return model.BaseExp
	}
	lvl := int64(level)
	return int64(difficulty)*lvl*lvl*lvl + NextLevelUp(level-1, difficulty)
}

// GiveNewLevel advances the character one level in their primary class.
// Original: GiveNewLevel() in Sly_CommonGuildUnit.cp
func GiveNewLevel(char *model.Character) {
	switch char.CharClass {
	case model.ClassFighter:
		char.FighterLvl++
		char.MaxHP += RandBetween(4, 5+char.FighterLvl)
		char.HitPoints = char.MaxHP
		char.Speed++
		char.Strength += RandBetween(1, 2)
		char.Dexterity += RandBetween(0, 1)
		char.Fame++
		char.Flirt1 += RandBetween(0, 2)
		char.Flirt2 += RandBetween(0, 2)

	case model.ClassMage:
		char.MageLvl++
		char.MaxHP += RandBetween(2, 3+char.MageLvl)
		char.HitPoints = char.MaxHP
		char.Speed++
		char.Strength += RandBetween(0, 1)
		char.Dexterity += RandBetween(0, 1)
		char.MaxPsyche += RandBetween(1, 2)
		char.Psyche = char.MaxPsyche
		char.Fame++
		char.Flirt1 += RandBetween(0, 2)
		char.Flirt2 += RandBetween(0, 2)

	case model.ClassThief:
		char.ThiefLvl++
		char.MaxHP += RandBetween(3, 4+char.ThiefLvl)
		char.HitPoints = char.MaxHP
		char.Speed += RandBetween(1, 2)
		char.Strength += RandBetween(1, 2)
		char.Dexterity += RandBetween(1, 2)
		char.Fame++
		char.Flirt1 += RandBetween(0, 2)
		char.Flirt2 += RandBetween(0, 2)
	}

	// Reset movement
	char.Movement = RandBetween(
		char.Speed-char.Speed/5,
		char.Speed+char.Speed/5,
	)
}

File: clients/web/mechanics/leveling.js
import { CharClass } from '../model/enums.js';
import {
  BASE_EXP, STARTING_COINS_HAND, STARTING_COINS_BANK,
  MAX_EXPLORATION, MAX_SPARS,
} from '../model/character.js';
import { randBetween, pick } from './rand.js';

// rollChar mirrors RollChar in internal/game/states/character_create.go.
// Sets level-1 stats and a starting weapon for the chosen class.
export function rollChar(c, charClass) {
  c.charClass = charClass;
  c.fighterLvl = 1;
  c.thiefLvl = 1;
  c.mageLvl = 1;
  c.totalExperience = BASE_EXP;
  c.spendingExperience = BASE_EXP;
  c.coinsHand = STARTING_COINS_HAND;
  c.coinsBank = STARTING_COINS_BANK;
  c.alive = true;
  c.fame = 0; c.honor = 0; c.faith = 0;
  c.flirt1 = 0; c.flirt2 = 0;
  c.totalFights = 0; c.fightsWon = 0;
  c.exploration = MAX_EXPLORATION;
  c.spars = MAX_SPARS;

  switch (charClass) {
    case CharClass.Fighter:
      c.maxHP = randBetween(15, 35);
      c.strength = randBetween(4, 12);
      c.dexterity = randBetween(4, 12);
      c.speed = randBetween(6, 12);
      c.maxPsyche = 1;
      c.movement = randBetween(8, 14);
      c.weapons[0] = { name: 'Short Sword', strike: 3, range: 0, actionStr: 'slash' };
      break;
    case CharClass.Thief:
      c.maxHP = randBetween(13, 26);
      c.strength = randBetween(3, 10);
      c.dexterity = randBetween(8, 20);
      c.speed = randBetween(10, 16);
      c.maxPsyche = 1;
      c.movement = randBetween(10, 18);
      c.weapons[0] = { name: 'Dagger', strike: 2, range: 1, actionStr: 'stab' };
      break;
    case CharClass.Mage:
      c.maxHP = randBetween(4, 17);
      c.strength = randBetween(2, 10);
      c.dexterity = randBetween(4, 14);
      c.speed = randBetween(6, 12);
      c.maxPsyche = randBetween(2, 4);
      c.movement = randBetween(6, 12);
      c.weapons[0] = { name: 'Staff', strike: 1, range: 0, actionStr: 'strike' };
      break;
  }

  c.hitPoints = c.maxHP;
  c.psyche = c.maxPsyche;
}

export function randomClass() {
  return pick([CharClass.Fighter, CharClass.Thief, CharClass.Mage]);
}
